File: common/src/main/java/com/alexstyl/specialdates/TimeOfDay.java
package com.alexstyl.specialdates;

import org.joda.time.LocalTime;

import static android.text.format.DateUtils.SECOND_IN_MILLIS;

public final class TimeOfDay {

    private static final long MINUTE_IN_MILLIS = SECOND_IN_MILLIS * 60;
    private static final long HOUR_IN_MILLIS = MINUTE_IN_MILLIS * 60;
    private static final String ZERO = "0";
    private static final String SEPARATOR = ":";
    private LocalTime dateTime;

    public static TimeOfDay now() {
        return new TimeOfDay(LocalTime.now());
    }

    private TimeOfDay(LocalTime dateTime) {
        this.dateTime = dateTime;
    }

    public TimeOfDay(int hour, int minute) {
        this(new LocalTime(hour, minute));
    }

    @Override
    public String toString() {
        int hour = getHours();
        StringBuilder str = new StringBuilder();
        if (isOneDigit(hour)) {
            str.append(ZERO);
        }
        str.append(hour)
                .append(SEPARATOR);
        int minute = getMinutes();
        if (isOneDigit(minute)) {
            str.append(ZERO);
        }
        str.append(minute);
        return str.toString();
    }

    private boolean isOneDigit(int value) {
        return value < 10;
    }

    public int getHours() {
        return dateTime.getHourOfDay();
    }

    public int getMinutes() {
        return dateTime.getMinuteOfHour();
    }

    public long toMillis() {
        return dateTime.getMillisOfDay();
    }

    public boolean isAfter(TimeOfDay timeOfDay) {
        return dateTime.isAfter(timeOfDay.dateTime);
    }
}

File: memento/src/main/java/com/alexstyl/specialdates/TimeOfDay.kt
package com.alexstyl.specialdates

import android.support.annotation.IntRange
import org.joda.time.LocalTime

data class TimeOfDay(private val dateTime: LocalTime) {

    val hours: Int
        get() = dateTime.hourOfDay

    val minutes: Int
        get() = dateTime.minuteOfHour

    override fun toString(): String {
        val hour = hours
        val str = StringBuilder()
        if (isOneDigit(hour)) {
            str.append(ZERO)
        }
        str.append(hour)
                .append(SEPARATOR)
        val minute = minutes
        if (isOneDigit(minute)) {
            str.append(ZERO)
        }
        str.append(minute)
        return str.toString()
    }

    private fun isOneDigit(value: Int): Boolean {
        return value < 10
    }

    fun toMillis(): Long {
        return dateTime.millisOfDay.toLong()
    }

    fun isAfter(other: TimeOfDay) = compareTo(other) > 0

    private fun compareTo(other: TimeOfDay): Int {
        return if (hours == other.hours) {
            minutes - other.minutes
        } else {
            hours - other.hours
        }
    }

    companion object {
        fun at(@IntRange(from = 0, to = 23) hour: Int, @IntRange(from = 0, to = 59) minute: Int): TimeOfDay {
            return TimeOfDay(LocalTime(hour, minute))
        }

        fun now(): TimeOfDay {
            return TimeOfDay(LocalTime.now())
        }

        private const val ZERO = "0"
        private const val SEPARATOR = ":"


    }
}
